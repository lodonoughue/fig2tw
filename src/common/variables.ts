export interface Collection {
  readonly name: string;
  readonly modes: string[];
  readonly defaultMode: string;
}

type Alias<V extends Variable> = Pick<V, "key">;

interface Color {
  readonly hex: string;
  readonly rgba: [number, number, number, number];
}

export interface Value<T extends string = string, V = unknown> {
  readonly type: T;
  readonly value: V;
}

export interface Variable<V extends Value = Value, S extends string = string> {
  readonly key: string;
  readonly name: string;
  readonly type: V["type"];
  readonly valuesByMode: Record<string, V | AliasValue>;
  readonly defaultValue: V;
  readonly collection: Collection;
  readonly scopes: S[];
}

export type AliasValue = Value<"alias", Alias<AnyVariable>>;

export type StringScope = "all-strings" | "font-family";
export type StringValue = Value<"string", string>;
export type StringVariable = Variable<StringValue, StringScope>;

// prettier-ignore
export type NumberScope = "all-numbers" | "radius" | "size" | "gap" | "stroke-width" | "font-size" | "line-height" | "letter-spacing" | "font-weight";
export type NumberValue = Value<"number", number>;
export type NumberVariable = Variable<NumberValue, NumberScope>;

export type BooleanScope = "all-booleans";
export type BooleanValue = Value<"boolean", boolean>;
export type BooleanVariable = Variable<BooleanValue, BooleanScope>;

// prettier-ignore
export type ColorScope = "all-colors" | "fill-color" | "stroke-color" | "text-color" | "effect-color";
export type ColorValue = Value<"color", Color>;
export type ColorVariable = Variable<ColorValue, ColorScope>;

export type AnyScope = StringScope | NumberScope | BooleanScope | ColorScope;

export type AnyValue =
  | AliasValue
  | StringValue
  | NumberValue
  | BooleanValue
  | ColorValue;

export type AnyVariable =
  | StringVariable
  | NumberVariable
  | BooleanVariable
  | ColorVariable;

export function isAliasValue(value: Value): value is AliasValue {
  return value.type === "alias";
}

export function isColorValue(value: Value): value is ColorValue {
  return value.type === "color";
}

export function isNumberValue(value: Value): value is NumberValue {
  return value.type === "number";
}

export function isStringValue(value: Value): value is StringValue {
  return value.type === "string";
}

export function isColorVariable(value: Variable): value is ColorVariable {
  return value.type === "color";
}

export function isNumberVariable(value: Variable): value is NumberVariable {
  return value.type === "number";
}

export function isBooleanVariable(value: Variable): value is BooleanVariable {
  return value.type === "boolean";
}

export function isStringVariable(value: Variable): value is StringVariable {
  return value.type === "string";
}

export function toColorHexPart(value: number): string {
  return Math.round(value * 255)
    .toString(16)
    .padStart(2, "0");
}

export function toColorRgbaPart(value: number): number {
  return Math.round(value * 255);
}
