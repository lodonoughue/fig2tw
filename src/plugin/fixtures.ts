/* v8 ignore start */
// Coverage is not calculated in fixtures, because it's only meant to help
// with the tests. It should  never be used in production code.

import { chain, zip } from "lodash";
import { vi } from "vitest";
import {
  isFigmaBooleanValue,
  isFigmaColorValue,
  isFigmaNumberValue,
  isFigmaStringValue,
} from "./figma";
import { fixtures as commonFixtures } from "@common/fixtures";

function createFigmaVariable({
  value,
  collection,
  id,
  key = "variable-key",
  name = "variable-name",
  description = "variable-description",
  hiddenFromPublishing = false,
  remote = false,
  variableCollectionId,
  resolvedType,
  valuesByMode,
  scopes = ["ALL_SCOPES"],
  codeSyntax = {},
  generateId = createIdGenerator(),
  getPublishStatusAsync = vi.fn(() => Promise.resolve("UNPUBLISHED" as const)),
  resolveForConsumer = vi.fn(),
  setValueForMode = vi.fn(),
  remove = vi.fn(),
  setVariableCodeSyntax = vi.fn(),
  removeVariableCodeSyntax = vi.fn(),
  getPluginData = vi.fn(),
  setPluginData = vi.fn(),
  getPluginDataKeys = vi.fn(),
  getSharedPluginData = vi.fn(),
  setSharedPluginData = vi.fn(),
  getSharedPluginDataKeys = vi.fn(),
}: FigmaVariableDefinition = {}): FigmaVariable {
  id ??= generateId();
  collection ??= inferVariableCollection(variableCollectionId, valuesByMode);
  variableCollectionId ??= inferVariableCollectionId(collection);
  value ??= inferValue(valuesByMode, resolvedType);
  resolvedType ??= inferResolvedType(value);
  valuesByMode ??= inferValuesByMode(value, collection);
  return {
    id,
    key,
    name,
    description,
    hiddenFromPublishing,
    remote,
    variableCollectionId,
    resolvedType,
    valuesByMode,
    scopes,
    codeSyntax,
    getPublishStatusAsync,
    resolveForConsumer,
    setValueForMode,
    remove,
    setVariableCodeSyntax,
    removeVariableCodeSyntax,
    getPluginData,
    setPluginData,
    getPluginDataKeys,
    getSharedPluginData,
    setSharedPluginData,
    getSharedPluginDataKeys,
  };
}

function createFigmaVariables(
  definitions: FigmaVariableDefinition[],
  generateId: () => string = createIdGenerator(),
): FigmaVariable[] {
  return definitions.map(it => createFigmaVariable({ ...it, generateId }));
}

function createFigmaCollection({
  id,
  key = "collection-key",
  name = "collection-name",
  hiddenFromPublishing = false,
  remote = false,
  modes,
  modeIds,
  modeNames,
  variableIds = [],
  defaultModeId,
  generateId = createIdGenerator(),
  getPublishStatusAsync = vi.fn(() => Promise.resolve("UNPUBLISHED" as const)),
  remove = vi.fn(),
  removeMode = vi.fn(),
  addMode = vi.fn(),
  renameMode = vi.fn(),
  getPluginData = vi.fn(),
  setPluginData = vi.fn(),
  getPluginDataKeys = vi.fn(),
  getSharedPluginData = vi.fn(),
  setSharedPluginData = vi.fn(),
  getSharedPluginDataKeys = vi.fn(),
}: FigmaCollectionDefinition = {}): FigmaCollection {
  id ??= generateId();
  modes ??= inferModes(modeIds, modeNames, defaultModeId, generateId);
  defaultModeId ??= inferDefaultModeId(modes);
  return {
    id,
    key,
    name,
    hiddenFromPublishing,
    remote,
    variableIds,
    modes,
    defaultModeId,
    getPublishStatusAsync,
    remove,
    removeMode,
    addMode,
    renameMode,
    getPluginData,
    setPluginData,
    getPluginDataKeys,
    getSharedPluginData,
    setSharedPluginData,
    getSharedPluginDataKeys,
  };
}

function createFigmaCollections(
  definitions: FigmaCollectionDefinition[],
  generateId: () => string = createIdGenerator(),
): FigmaCollection[] {
  return definitions.map(it => createFigmaCollection({ ...it, generateId }));
}

function createFigmaMode({
  modeId,
  name,
  generateId = createIdGenerator(),
}: FigmaModeDefinition = {}): FigmaMode {
  modeId ??= generateId ? generateId() : "mode-id";
  name ??= `${modeId}-name`;
  return { modeId, name };
}

function createFigmaRgbColor([
  r = 1,
  g = 1,
  b = 1,
]: FigmaColorDefinition = []): RGB {
  return { r, g, b };
}

function createFigmaRgbaColor([
  r = 1,
  g = 1,
  b = 1,
  a = 1,
]: FigmaColorDefinition = []): RGBA {
  return { r, g, b, a };
}

function createFigmaVariableAlias({
  id = "alias-id",
}: FigmaVariableAliasDefinition = {}): VariableAlias {
  return { type: "VARIABLE_ALIAS", id };
}

function createIdGenerator() {
  let id = 0;
  return () => `id-${id++}`;
}

function inferModes(
  modeIds: FigmaCollectionDefinition["modeIds"],
  modeNames: FigmaCollectionDefinition["modeNames"],
  defaultModeId: FigmaCollectionDefinition["defaultModeId"],
  generateId: FigmaCollectionDefinition["generateId"],
): FigmaCollection["modes"] {
  if (modeNames != null && modeIds != null) {
    return zip(modeIds, modeNames).map(([modeId, name]) =>
      createFigmaMode({ modeId, name }),
    );
  }

  if (modeIds != null) {
    return modeIds.map(modeId => createFigmaMode({ modeId }));
  }

  if (modeNames != null) {
    return modeNames.map((name, index) =>
      createFigmaMode({
        modeId: index === 0 && defaultModeId ? defaultModeId : undefined,
        name,
        generateId,
      }),
    );
  }

  return [createFigmaMode({ modeId: defaultModeId })];
}

function inferDefaultModeId(
  modes?: FigmaCollectionDefinition["modes"],
): FigmaCollection["defaultModeId"] {
  if (modes != null && modes.length) {
    return modes[0].modeId;
  }
  return createFigmaMode().modeId;
}

function inferVariableCollectionId(
  collection: FigmaVariableDefinition["collection"],
) {
  return collection ? collection.id : createFigmaCollection().id;
}

function inferValue(
  valuesByMode: FigmaVariableDefinition["valuesByMode"],
  resolvedType: FigmaVariableDefinition["resolvedType"],
): FigmaValue {
  const defaultValue = valuesByMode
    ? Object.values(valuesByMode)[0]
    : undefined;

  if (defaultValue) return defaultValue;
  if (resolvedType === "FLOAT") return 42;
  if (resolvedType === "BOOLEAN") return true;
  if (resolvedType === "COLOR") return createFigmaRgbaColor();

  return "foo";
}

function inferValuesByMode(value: FigmaValue, collection: FigmaCollection) {
  return chain(collection.modes)
    .keyBy("modeId")
    .mapValues(() => value)
    .value();
}

function inferResolvedType(value: FigmaValue) {
  if (isFigmaNumberValue(value)) return "FLOAT";
  if (isFigmaBooleanValue(value)) return "BOOLEAN";
  if (isFigmaStringValue(value)) return "STRING";
  if (isFigmaColorValue(value)) return "COLOR";
  return "STRING";
}

function inferVariableCollection(
  collectionId: FigmaVariableDefinition["variableCollectionId"],
  valuesByMode: FigmaVariableDefinition["valuesByMode"],
) {
  const modes = valuesByMode
    ? Object.keys(valuesByMode).map(modeId => createFigmaMode({ modeId }))
    : undefined;

  return createFigmaCollection({ id: collectionId, modes });
}

export const fixtures = {
  ...commonFixtures,
  createFigmaVariable,
  createFigmaVariables,
  createFigmaCollection,
  createFigmaCollections,
  createFigmaMode,
  createFigmaRgbColor,
  createFigmaRgbaColor,
  createFigmaVariableAlias,
  createIdGenerator,
};

type FigmaMode = VariableCollection["modes"][number];
type FigmaCollection = VariableCollection;
type FigmaVariable = Variable;
type FigmaValue = Variable["valuesByMode"][string];
type FigmaColorDefinition = [number, number, number, number?] | [];

interface FigmaVariableAliasDefinition {
  id?: string;
}

interface FigmaModeDefinition extends Partial<FigmaMode> {
  generateId?: () => string;
}

interface FigmaCollectionDefinition extends Partial<FigmaCollection> {
  modeNames?: string[];
  modeIds?: string[];
  generateId?: () => string;
}

interface FigmaVariableDefinition extends Partial<FigmaVariable> {
  value?: FigmaValue;
  collection?: FigmaCollection;
  generateId?: () => string;
}
