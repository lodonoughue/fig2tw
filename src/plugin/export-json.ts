import { Config } from "@common/config";
import { loadVariables } from "./variables";
import { AnyVariable } from "@common/variables";
import { chain } from "lodash";

export async function exportJson(config: Config) {
  const variables = await loadVariables();
  return JSON.stringify(convertToJson(variables), null, config.tabWidth);
}

function convertToJson(variables: AnyVariable[]) {
  return chain(variables).keyBy("key").value();
}
