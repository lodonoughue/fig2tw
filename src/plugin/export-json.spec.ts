import { fixtures } from "@common/fixtures";
import { describe, expect, it, vi } from "vitest";
import { loadVariables } from "./variables";
import { exportJson } from "./export-json";

vi.mock("@plugin/variables", async importOriginal => {
  const original = await importOriginal<typeof import("@plugin/variables")>();
  vi.spyOn(original, "loadVariables");
  return original;
});

describe.each([
  { config: fixtures.createConfig({ tabWidth: 4 }) },
  { config: fixtures.createConfig({ tabWidth: 2 }) },
])("exportJson (tabWidth=$config.tabWidth)", ({ config }) => {
  it("should export color variables", async () => {
    vi.mocked(loadVariables).mockResolvedValue([
      fixtures.createColorVariable({
        key: "Color/Primary",
        valuesByMode: {
          light: fixtures.createColorValue({ value: { hex: "#800000" } }),
          dark: fixtures.createColorValue({ value: { hex: "#c00000" } }),
        },
      }),
      fixtures.createColorVariable({
        key: "Color/Secondary",
        valuesByMode: {
          light: fixtures.createColorValue({ value: { hex: "#008000" } }),
          dark: fixtures.createColorValue({ value: { hex: "#00c000" } }),
        },
      }),
      fixtures.createColorVariable({
        key: "Color/Tertiary",
        valuesByMode: {
          light: fixtures.createColorValue({ value: { hex: "#000080" } }),
          dark: fixtures.createColorValue({ value: { hex: "#0000c0" } }),
        },
      }),
      fixtures.createColorVariable({
        key: "Color/Alias",
        valuesByMode: {
          light: fixtures.createAliasValue({ value: { key: "Color/Primary" } }),
          dark: fixtures.createAliasValue({
            value: { key: "Color/Secondary" },
          }),
        },
      }),
    ]);

    const result = await exportJson(config);
    expect(result).toMatchSnapshot();
  });

  it("should export number variables", async () => {
    vi.mocked(loadVariables).mockResolvedValue([
      fixtures.createNumberVariable({
        key: "Number/Gap",
        scopes: ["gap"],
        valuesByMode: {
          regular: fixtures.createNumberValue({ value: 8 }),
          condensed: fixtures.createNumberValue({ value: 4 }),
        },
      }),
      fixtures.createNumberVariable({
        key: "Number/Radius",
        scopes: ["radius"],
        valuesByMode: {
          regular: fixtures.createNumberValue({ value: 16 }),
          condensed: fixtures.createNumberValue({ value: 12 }),
        },
      }),
      fixtures.createNumberVariable({
        key: "Number/Size",
        scopes: ["size"],
        valuesByMode: {
          regular: fixtures.createNumberValue({ value: 20 }),
          condensed: fixtures.createNumberValue({ value: 24 }),
        },
      }),
      fixtures.createNumberVariable({
        key: "Number/Alias",
        valuesByMode: {
          regular: fixtures.createAliasValue({
            value: { key: "Number/Radius" },
          }),
          condensed: fixtures.createAliasValue({
            value: { key: "Number/Size" },
          }),
        },
      }),
    ]);

    const result = await exportJson(config);
    expect(result).toMatchSnapshot();
  });

  it("should export string variables", async () => {
    vi.mocked(loadVariables).mockResolvedValue([
      fixtures.createStringVariable({
        key: "String/Font Family",
        scopes: ["font-family"],
        valuesByMode: {
          regular: fixtures.createStringValue({ value: "Arial" }),
          comic: fixtures.createStringValue({ value: "Comic Sans" }),
        },
      }),
      fixtures.createStringVariable({
        key: "String/Alias",
        valuesByMode: {
          regular: fixtures.createAliasValue({
            value: { key: "String/Target" },
          }),
          comic: fixtures.createAliasValue({ value: { key: "String/Target" } }),
        },
      }),
    ]);

    const result = await exportJson(config);
    expect(result).toMatchSnapshot();
  });

  it("should export boolean variables", async () => {
    vi.mocked(loadVariables).mockResolvedValue([
      fixtures.createBooleanVariable({
        key: "Boolean/Falsy",
        scopes: ["all-booleans"],
        valuesByMode: {
          true: fixtures.createBooleanValue({ value: false }),
          false: fixtures.createBooleanValue({ value: true }),
        },
      }),
      fixtures.createBooleanVariable({
        key: "Boolean/Alias",
        valuesByMode: {
          regular: fixtures.createAliasValue({
            value: { key: "Boolean/Target" },
          }),
          comic: fixtures.createAliasValue({
            value: { key: "Boolean/Target" },
          }),
        },
      }),
    ]);

    const result = await exportJson(config);
    expect(result).toMatchSnapshot();
  });
});
