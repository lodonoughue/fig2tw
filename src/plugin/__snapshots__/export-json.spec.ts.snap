// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`exportJson (tabWidth=2) > should export boolean variables 1`] = `
"{
  "Boolean/Falsy": {
    "type": "boolean",
    "key": "Boolean/Falsy",
    "name": "Falsy",
    "valuesByMode": {
      "true": {
        "type": "boolean",
        "value": false
      },
      "false": {
        "type": "boolean",
        "value": true
      }
    },
    "defaultValue": {
      "type": "boolean",
      "value": false
    },
    "collection": {
      "name": "Boolean",
      "modes": [
        "true",
        "false"
      ],
      "defaultMode": "true"
    },
    "scopes": [
      "all-booleans"
    ]
  },
  "Boolean/Alias": {
    "type": "boolean",
    "key": "Boolean/Alias",
    "name": "Alias",
    "valuesByMode": {
      "regular": {
        "type": "alias",
        "value": {
          "key": "Boolean/Target"
        }
      },
      "comic": {
        "type": "alias",
        "value": {
          "key": "Boolean/Target"
        }
      }
    },
    "defaultValue": {
      "type": "boolean",
      "value": true
    },
    "collection": {
      "name": "Boolean",
      "modes": [
        "regular",
        "comic"
      ],
      "defaultMode": "regular"
    },
    "scopes": [
      "all-booleans"
    ]
  }
}"
`;

exports[`exportJson (tabWidth=2) > should export color variables 1`] = `
"{
  "Color/Primary": {
    "type": "color",
    "key": "Color/Primary",
    "name": "Primary",
    "valuesByMode": {
      "light": {
        "type": "color",
        "value": {
          "hex": "#800000",
          "rgba": [
            128,
            0,
            0,
            1
          ]
        }
      },
      "dark": {
        "type": "color",
        "value": {
          "hex": "#c00000",
          "rgba": [
            192,
            0,
            0,
            1
          ]
        }
      }
    },
    "defaultValue": {
      "type": "color",
      "value": {
        "hex": "#800000",
        "rgba": [
          128,
          0,
          0,
          1
        ]
      }
    },
    "collection": {
      "name": "Color",
      "modes": [
        "light",
        "dark"
      ],
      "defaultMode": "light"
    },
    "scopes": [
      "all-colors"
    ]
  },
  "Color/Secondary": {
    "type": "color",
    "key": "Color/Secondary",
    "name": "Secondary",
    "valuesByMode": {
      "light": {
        "type": "color",
        "value": {
          "hex": "#008000",
          "rgba": [
            0,
            128,
            0,
            1
          ]
        }
      },
      "dark": {
        "type": "color",
        "value": {
          "hex": "#00c000",
          "rgba": [
            0,
            192,
            0,
            1
          ]
        }
      }
    },
    "defaultValue": {
      "type": "color",
      "value": {
        "hex": "#008000",
        "rgba": [
          0,
          128,
          0,
          1
        ]
      }
    },
    "collection": {
      "name": "Color",
      "modes": [
        "light",
        "dark"
      ],
      "defaultMode": "light"
    },
    "scopes": [
      "all-colors"
    ]
  },
  "Color/Tertiary": {
    "type": "color",
    "key": "Color/Tertiary",
    "name": "Tertiary",
    "valuesByMode": {
      "light": {
        "type": "color",
        "value": {
          "hex": "#000080",
          "rgba": [
            0,
            0,
            128,
            1
          ]
        }
      },
      "dark": {
        "type": "color",
        "value": {
          "hex": "#0000c0",
          "rgba": [
            0,
            0,
            192,
            1
          ]
        }
      }
    },
    "defaultValue": {
      "type": "color",
      "value": {
        "hex": "#000080",
        "rgba": [
          0,
          0,
          128,
          1
        ]
      }
    },
    "collection": {
      "name": "Color",
      "modes": [
        "light",
        "dark"
      ],
      "defaultMode": "light"
    },
    "scopes": [
      "all-colors"
    ]
  },
  "Color/Alias": {
    "type": "color",
    "key": "Color/Alias",
    "name": "Alias",
    "valuesByMode": {
      "light": {
        "type": "alias",
        "value": {
          "key": "Color/Primary"
        }
      },
      "dark": {
        "type": "alias",
        "value": {
          "key": "Color/Secondary"
        }
      }
    },
    "defaultValue": {
      "type": "color",
      "value": {
        "hex": "#ff000000",
        "rgba": [
          255,
          0,
          0,
          0
        ]
      }
    },
    "collection": {
      "name": "Color",
      "modes": [
        "light",
        "dark"
      ],
      "defaultMode": "light"
    },
    "scopes": [
      "all-colors"
    ]
  }
}"
`;

exports[`exportJson (tabWidth=2) > should export number variables 1`] = `
"{
  "Number/Gap": {
    "type": "number",
    "key": "Number/Gap",
    "name": "Gap",
    "valuesByMode": {
      "regular": {
        "type": "number",
        "value": 8
      },
      "condensed": {
        "type": "number",
        "value": 4
      }
    },
    "defaultValue": {
      "type": "number",
      "value": 8
    },
    "collection": {
      "name": "Number",
      "modes": [
        "regular",
        "condensed"
      ],
      "defaultMode": "regular"
    },
    "scopes": [
      "gap"
    ]
  },
  "Number/Radius": {
    "type": "number",
    "key": "Number/Radius",
    "name": "Radius",
    "valuesByMode": {
      "regular": {
        "type": "number",
        "value": 16
      },
      "condensed": {
        "type": "number",
        "value": 12
      }
    },
    "defaultValue": {
      "type": "number",
      "value": 16
    },
    "collection": {
      "name": "Number",
      "modes": [
        "regular",
        "condensed"
      ],
      "defaultMode": "regular"
    },
    "scopes": [
      "radius"
    ]
  },
  "Number/Size": {
    "type": "number",
    "key": "Number/Size",
    "name": "Size",
    "valuesByMode": {
      "regular": {
        "type": "number",
        "value": 20
      },
      "condensed": {
        "type": "number",
        "value": 24
      }
    },
    "defaultValue": {
      "type": "number",
      "value": 20
    },
    "collection": {
      "name": "Number",
      "modes": [
        "regular",
        "condensed"
      ],
      "defaultMode": "regular"
    },
    "scopes": [
      "size"
    ]
  },
  "Number/Alias": {
    "type": "number",
    "key": "Number/Alias",
    "name": "Alias",
    "valuesByMode": {
      "regular": {
        "type": "alias",
        "value": {
          "key": "Number/Radius"
        }
      },
      "condensed": {
        "type": "alias",
        "value": {
          "key": "Number/Size"
        }
      }
    },
    "defaultValue": {
      "type": "number",
      "value": 42
    },
    "collection": {
      "name": "Number",
      "modes": [
        "regular",
        "condensed"
      ],
      "defaultMode": "regular"
    },
    "scopes": [
      "all-numbers"
    ]
  }
}"
`;

exports[`exportJson (tabWidth=2) > should export string variables 1`] = `
"{
  "String/Font Family": {
    "type": "string",
    "key": "String/Font Family",
    "name": "Font Family",
    "valuesByMode": {
      "regular": {
        "type": "string",
        "value": "Arial"
      },
      "comic": {
        "type": "string",
        "value": "Comic Sans"
      }
    },
    "defaultValue": {
      "type": "string",
      "value": "Arial"
    },
    "collection": {
      "name": "String",
      "modes": [
        "regular",
        "comic"
      ],
      "defaultMode": "regular"
    },
    "scopes": [
      "font-family"
    ]
  },
  "String/Alias": {
    "type": "string",
    "key": "String/Alias",
    "name": "Alias",
    "valuesByMode": {
      "regular": {
        "type": "alias",
        "value": {
          "key": "String/Target"
        }
      },
      "comic": {
        "type": "alias",
        "value": {
          "key": "String/Target"
        }
      }
    },
    "defaultValue": {
      "type": "string",
      "value": "foo"
    },
    "collection": {
      "name": "String",
      "modes": [
        "regular",
        "comic"
      ],
      "defaultMode": "regular"
    },
    "scopes": [
      "all-strings"
    ]
  }
}"
`;

exports[`exportJson (tabWidth=4) > should export boolean variables 1`] = `
"{
    "Boolean/Falsy": {
        "type": "boolean",
        "key": "Boolean/Falsy",
        "name": "Falsy",
        "valuesByMode": {
            "true": {
                "type": "boolean",
                "value": false
            },
            "false": {
                "type": "boolean",
                "value": true
            }
        },
        "defaultValue": {
            "type": "boolean",
            "value": false
        },
        "collection": {
            "name": "Boolean",
            "modes": [
                "true",
                "false"
            ],
            "defaultMode": "true"
        },
        "scopes": [
            "all-booleans"
        ]
    },
    "Boolean/Alias": {
        "type": "boolean",
        "key": "Boolean/Alias",
        "name": "Alias",
        "valuesByMode": {
            "regular": {
                "type": "alias",
                "value": {
                    "key": "Boolean/Target"
                }
            },
            "comic": {
                "type": "alias",
                "value": {
                    "key": "Boolean/Target"
                }
            }
        },
        "defaultValue": {
            "type": "boolean",
            "value": true
        },
        "collection": {
            "name": "Boolean",
            "modes": [
                "regular",
                "comic"
            ],
            "defaultMode": "regular"
        },
        "scopes": [
            "all-booleans"
        ]
    }
}"
`;

exports[`exportJson (tabWidth=4) > should export color variables 1`] = `
"{
    "Color/Primary": {
        "type": "color",
        "key": "Color/Primary",
        "name": "Primary",
        "valuesByMode": {
            "light": {
                "type": "color",
                "value": {
                    "hex": "#800000",
                    "rgba": [
                        128,
                        0,
                        0,
                        1
                    ]
                }
            },
            "dark": {
                "type": "color",
                "value": {
                    "hex": "#c00000",
                    "rgba": [
                        192,
                        0,
                        0,
                        1
                    ]
                }
            }
        },
        "defaultValue": {
            "type": "color",
            "value": {
                "hex": "#800000",
                "rgba": [
                    128,
                    0,
                    0,
                    1
                ]
            }
        },
        "collection": {
            "name": "Color",
            "modes": [
                "light",
                "dark"
            ],
            "defaultMode": "light"
        },
        "scopes": [
            "all-colors"
        ]
    },
    "Color/Secondary": {
        "type": "color",
        "key": "Color/Secondary",
        "name": "Secondary",
        "valuesByMode": {
            "light": {
                "type": "color",
                "value": {
                    "hex": "#008000",
                    "rgba": [
                        0,
                        128,
                        0,
                        1
                    ]
                }
            },
            "dark": {
                "type": "color",
                "value": {
                    "hex": "#00c000",
                    "rgba": [
                        0,
                        192,
                        0,
                        1
                    ]
                }
            }
        },
        "defaultValue": {
            "type": "color",
            "value": {
                "hex": "#008000",
                "rgba": [
                    0,
                    128,
                    0,
                    1
                ]
            }
        },
        "collection": {
            "name": "Color",
            "modes": [
                "light",
                "dark"
            ],
            "defaultMode": "light"
        },
        "scopes": [
            "all-colors"
        ]
    },
    "Color/Tertiary": {
        "type": "color",
        "key": "Color/Tertiary",
        "name": "Tertiary",
        "valuesByMode": {
            "light": {
                "type": "color",
                "value": {
                    "hex": "#000080",
                    "rgba": [
                        0,
                        0,
                        128,
                        1
                    ]
                }
            },
            "dark": {
                "type": "color",
                "value": {
                    "hex": "#0000c0",
                    "rgba": [
                        0,
                        0,
                        192,
                        1
                    ]
                }
            }
        },
        "defaultValue": {
            "type": "color",
            "value": {
                "hex": "#000080",
                "rgba": [
                    0,
                    0,
                    128,
                    1
                ]
            }
        },
        "collection": {
            "name": "Color",
            "modes": [
                "light",
                "dark"
            ],
            "defaultMode": "light"
        },
        "scopes": [
            "all-colors"
        ]
    },
    "Color/Alias": {
        "type": "color",
        "key": "Color/Alias",
        "name": "Alias",
        "valuesByMode": {
            "light": {
                "type": "alias",
                "value": {
                    "key": "Color/Primary"
                }
            },
            "dark": {
                "type": "alias",
                "value": {
                    "key": "Color/Secondary"
                }
            }
        },
        "defaultValue": {
            "type": "color",
            "value": {
                "hex": "#ff000000",
                "rgba": [
                    255,
                    0,
                    0,
                    0
                ]
            }
        },
        "collection": {
            "name": "Color",
            "modes": [
                "light",
                "dark"
            ],
            "defaultMode": "light"
        },
        "scopes": [
            "all-colors"
        ]
    }
}"
`;

exports[`exportJson (tabWidth=4) > should export number variables 1`] = `
"{
    "Number/Gap": {
        "type": "number",
        "key": "Number/Gap",
        "name": "Gap",
        "valuesByMode": {
            "regular": {
                "type": "number",
                "value": 8
            },
            "condensed": {
                "type": "number",
                "value": 4
            }
        },
        "defaultValue": {
            "type": "number",
            "value": 8
        },
        "collection": {
            "name": "Number",
            "modes": [
                "regular",
                "condensed"
            ],
            "defaultMode": "regular"
        },
        "scopes": [
            "gap"
        ]
    },
    "Number/Radius": {
        "type": "number",
        "key": "Number/Radius",
        "name": "Radius",
        "valuesByMode": {
            "regular": {
                "type": "number",
                "value": 16
            },
            "condensed": {
                "type": "number",
                "value": 12
            }
        },
        "defaultValue": {
            "type": "number",
            "value": 16
        },
        "collection": {
            "name": "Number",
            "modes": [
                "regular",
                "condensed"
            ],
            "defaultMode": "regular"
        },
        "scopes": [
            "radius"
        ]
    },
    "Number/Size": {
        "type": "number",
        "key": "Number/Size",
        "name": "Size",
        "valuesByMode": {
            "regular": {
                "type": "number",
                "value": 20
            },
            "condensed": {
                "type": "number",
                "value": 24
            }
        },
        "defaultValue": {
            "type": "number",
            "value": 20
        },
        "collection": {
            "name": "Number",
            "modes": [
                "regular",
                "condensed"
            ],
            "defaultMode": "regular"
        },
        "scopes": [
            "size"
        ]
    },
    "Number/Alias": {
        "type": "number",
        "key": "Number/Alias",
        "name": "Alias",
        "valuesByMode": {
            "regular": {
                "type": "alias",
                "value": {
                    "key": "Number/Radius"
                }
            },
            "condensed": {
                "type": "alias",
                "value": {
                    "key": "Number/Size"
                }
            }
        },
        "defaultValue": {
            "type": "number",
            "value": 42
        },
        "collection": {
            "name": "Number",
            "modes": [
                "regular",
                "condensed"
            ],
            "defaultMode": "regular"
        },
        "scopes": [
            "all-numbers"
        ]
    }
}"
`;

exports[`exportJson (tabWidth=4) > should export string variables 1`] = `
"{
    "String/Font Family": {
        "type": "string",
        "key": "String/Font Family",
        "name": "Font Family",
        "valuesByMode": {
            "regular": {
                "type": "string",
                "value": "Arial"
            },
            "comic": {
                "type": "string",
                "value": "Comic Sans"
            }
        },
        "defaultValue": {
            "type": "string",
            "value": "Arial"
        },
        "collection": {
            "name": "String",
            "modes": [
                "regular",
                "comic"
            ],
            "defaultMode": "regular"
        },
        "scopes": [
            "font-family"
        ]
    },
    "String/Alias": {
        "type": "string",
        "key": "String/Alias",
        "name": "Alias",
        "valuesByMode": {
            "regular": {
                "type": "alias",
                "value": {
                    "key": "String/Target"
                }
            },
            "comic": {
                "type": "alias",
                "value": {
                    "key": "String/Target"
                }
            }
        },
        "defaultValue": {
            "type": "string",
            "value": "foo"
        },
        "collection": {
            "name": "String",
            "modes": [
                "regular",
                "comic"
            ],
            "defaultMode": "regular"
        },
        "scopes": [
            "all-strings"
        ]
    }
}"
`;
